service/lib/y2storage/proposal/agama_device_planner.rb

```diff
@@ -0,0 +1,124 @@
+        @config = config
+        @issues_list = issues_list
+      end
+
+      # Planned devices according to the given settings.
+      #
+      # @return [Array] Array of planned devices.
+      def planned_devices(_setting)
+        raise NotImplementedError
+      end
+
+      private
+
+      # @param planned [Planned::Disk, Planned::Partition]
+      # @param settings [#format, #mount]
+      def configure_device(planned, settings)
+        # TODO configure_encrypt
+        configure_format(planned, settings.format) if settings.format
+        configure_mount(planned, settings.mount) if settings.mount
+      end
+
+      # @param planned [Planned::Disk, Planned::Partition]
+      # @param settings [Agama::Storage::Settings::Format]
+      def configure_format(planned, settings)
+        planned.label = settings.label
+        planned.mkfs_options = settings.mkfs_options
+        configure_filesystem(planned, settings.filesystem) if settings.filesystem
+      end
+
+      # @param planned [Planned::Disk, Planned::Partition]
+      # @param settings [Agama::Storage::Settings::Filesystem]
+      def configure_filesystem(planned, settings)
+        planned.filesystem_type = settings.type
+        configure_btrfs(planned, settings.btrfs) if settings.btrfs
+      end
+
+      # @param planned [Planned::Disk, Planned::Partition]
+      # @param settings [Agama::Storage::Settings::Btrfs]
+      def configure_btrfs(planned, settings)
+        planned.snapshots = settings.snapshots?
+        planned.default_subvolume = settings.default_subvolume
+        planned.subvolumes = settings.subvolumes
+      end
+
+      # @param planned [Planned::Disk, Planned::Partition]
+      # @param settings [Agama::Storage::Settings::Mount]
+      def configure_mount(planned, settings)
+        planned.mount_point = settings.path
+        planned.mount_by = settings.mount_by
+        planned.fstab_options = settings.options
+        # FIXME: Is this needed? Or #options is enough?
+        # planned.read_only = settings.read_only?
+      end
+
+      # @param planned [Planned::Partition]
+      # @param settings [Agama::Storage::Settings::Size]
+      def configure_size(planned, settings)
+        planned.min_size = settings.min
+        planned.max_size = settings.max
+      end
+
+      # @param planned [Planned::Disk]
+      # @param settings [Agama::Storage::Settings::Drive]
+      def configure_partitions(planned, settings)
+        planned.partitions = settings.partitions.map do |partition_settings|
+          planned_partition(partition_settings).tap { |p| p.disk = settings.found_device.name }
+        end
+      end
+
+      # @param settings [Agama::Storage::Settings::Partition]
+      # @return [Planned::Partition]
+      def planned_partition(settings)
+        Planned::Partition.new(nil, nil).tap do |planned|
+          planned.partition_id = settings.id
+          # FIXME: do we want to allow to force primary partitions? That was not really
```

Comment:
I would remove `Partition.type` (string) from the schema and would totally let the algorithm decide whether a partition should be primary or logical.

Reply:
Removed.